File: src/docforge_sdk/docforge_sdk/models/explorer.py
# ====== Code Summary ======
# Response/request models for the document explorer resource, mirrored field-for-field from the
# DocForge backend router models: the catalogue list item, the per-document detail + resolved
# metadata, a page's render reference, a chunk with its composition, and the chunk-toggle contract.

# ====== Standard Library Imports ======
from datetime import datetime
from typing import Any

# ====== Third-Party Library Imports ======
from pydantic import BaseModel, Field

# ====== Local Project Imports ======
from ._shared import DocumentStatus, FieldOrigin, SourceKind


class MetadataValue(BaseModel):
    """
    One resolved metadata value — the field name, the stored value and the origin that filled it.

    Attributes:
        field_name (str): The schema field this value fills.
        value (Any): The stored JSON value (typed by the field).
        origin (FieldOrigin): Who filled it: user / system / generated.
    """

    field_name: str = Field(description="The schema field this value fills.")
    value: Any = Field(description="The stored JSON value (typed by the field).")
    origin: FieldOrigin = Field(description="Who filled it: user / system / generated.")


class DocumentListItem(BaseModel):
    """
    One row of a collection's document catalogue (the browse list).

    Attributes:
        id (str): The document's UUID.
        filename (str): The uploaded file name.
        format (str): The file extension (pdf, docx, …).
        status (DocumentStatus): Ingestion lifecycle state.
        page_count (int | None): Pages once parsed (None before).
        file_size (int): Original size in bytes.
        created_at (datetime | None): Admission timestamp.
        title (str): Learned title ('' before parse).
        language (str | None): Detected language (None before parse).
        enabled (bool): Document-level searchability toggle.
    """

    id: str = Field(description="The document's UUID.")
    filename: str = Field(description="The uploaded file name.")
    format: str = Field(description="The file extension (pdf, docx, …).")
    status: DocumentStatus = Field(description="Ingestion lifecycle state.")
    page_count: int | None = Field(default=None, description="Pages once parsed (None before).")
    file_size: int = Field(description="Original size in bytes.")
    created_at: datetime | None = Field(default=None, description="Admission timestamp.")
    title: str = Field(description="Learned title ('' before parse).")
    language: str | None = Field(default=None, description="Detected language (None before parse).")
    enabled: bool = Field(description="Document-level searchability toggle.")


class DocumentDetail(BaseModel):
    """
    A document's full facts plus its resolved document-level metadata.

    Attributes:
        id (str): The document's UUID.
        collection_id (str): Owning collection.
        filename (str): The uploaded file name.
        format (str): The file extension.
        mime_type (str): The upload's declared content type.
        file_size (int): Original size in bytes.
        page_count (int | None): Pages once parsed.
        language (str | None): Detected document language.
        title (str): Learned title ('' before parse).
        source_kind (SourceKind): Acquisition routing.
        status (DocumentStatus): Ingestion lifecycle state.
        source_hash (str): Content address of the original bytes (blob key).
        pdf_blob_hash (str | None): Canonical PDF view blob (or None).
        simhash (str | None): Near-duplicate signature (or None).
        pipeline_version (str): Pipeline config identity the run used.
        created_at (datetime | None): Admission timestamp.
        enabled (bool): Document-level searchability toggle.
        metadata (list[MetadataValue]): Document-level values (declared and generated).
    """

    id: str = Field(description="The document's UUID.")
    collection_id: str = Field(description="Owning collection.")
    filename: str = Field(description="The uploaded file name.")
    format: str = Field(description="The file extension.")
    mime_type: str = Field(description="The upload's declared content type.")
    file_size: int = Field(description="Original size in bytes.")
    page_count: int | None = Field(default=None, description="Pages once parsed.")
    language: str | None = Field(default=None, description="Detected document language.")
    title: str = Field(description="Learned title ('' before parse).")
    source_kind: SourceKind = Field(description="Acquisition routing.")
    status: DocumentStatus = Field(description="Ingestion lifecycle state.")
    source_hash: str = Field(description="Content address of the original bytes (blob key).")
    pdf_blob_hash: str | None = Field(
        default=None, description="Canonical PDF view blob (or None)."
    )
    simhash: str | None = Field(default=None, description="Near-duplicate signature (or None).")
    pipeline_version: str = Field(description="Pipeline config identity the run used.")
    created_at: datetime | None = Field(default=None, description="Admission timestamp.")
    enabled: bool = Field(description="Document-level searchability toggle.")
    metadata: list[MetadataValue] = Field(
        default_factory=list, description="Document-level values (declared and generated)."
    )


class PageInfo(BaseModel):
    """
    One page's geometry, routing and its render blob reference.

    Attributes:
        page_number (int): 0-based page index (first page = 0).
        width (float | None): Page width in points (None if unknown).
        height (float | None): Page height in points (None if unknown).
        is_scanned (bool): Whether this page was routed as image-only (OCR/VLM).
        language (str | None): Per-page detected language.
        render_blob_hash (str | None): Blob of the rasterized page (None when render was off).
    """

    page_number: int = Field(description="0-based page index (first page = 0).")
    width: float | None = Field(default=None, description="Page width in points (None if unknown).")
    height: float | None = Field(
        default=None, description="Page height in points (None if unknown)."
    )
    is_scanned: bool = Field(description="Whether this page was routed as image-only (OCR/VLM).")
    language: str | None = Field(default=None, description="Per-page detected language.")
    render_blob_hash: str | None = Field(
        default=None, description="Blob of the rasterized page (None when render was off)."
    )


class ChunkInfo(BaseModel):
    """
    One retrieval chunk — enriched text, composition (block ids) and generated metadata.

    Attributes:
        id (str): The chunk UUID (doubles as the Qdrant point id).
        chunk_index (int): Position of the chunk within the document.
        text (str): The enriched, embedded text.
        token_count (int): Token length of the enriched text.
        is_indexed (bool): Whether the chunk is upserted into Qdrant.
        role (str): Structural classification set by the pipeline.
        enabled (bool): Effective searchability (override when set, else the role default).
        strategy (str): The chunking strategy that produced it.
        parent_id (str | None): Parent chunk (hierarchical chunking).
        block_ids (list[str]): Composing IR block ids, in assembly order.
        metadata (list[MetadataValue]): Per-chunk generated metadata values.
    """

    id: str = Field(description="The chunk UUID (doubles as the Qdrant point id).")
    chunk_index: int = Field(description="Position of the chunk within the document.")
    text: str = Field(description="The enriched, embedded text.")
    token_count: int = Field(description="Token length of the enriched text.")
    is_indexed: bool = Field(description="Whether the chunk is upserted into Qdrant.")
    role: str = Field(description="Structural classification set by the pipeline.")
    enabled: bool = Field(description="Effective searchability (override or role default).")
    strategy: str = Field(description="The chunking strategy that produced it.")
    parent_id: str | None = Field(default=None, description="Parent chunk (hierarchical chunking).")
    block_ids: list[str] = Field(
        default_factory=list, description="Composing IR block ids, in assembly (position) order."
    )
    metadata: list[MetadataValue] = Field(
        default_factory=list, description="Per-chunk generated metadata values."
    )


class ChunkEnabledPatch(BaseModel):
    """
    The desired searchability state for one chunk (its enabled_override).

    Attributes:
        enabled (bool): True to make the chunk searchable, False to hide it.
    """

    enabled: bool = Field(description="True to make the chunk searchable, False to hide it.")


class BulkChunkEnabledPatch(BaseModel):
    """
    Toggle several chunks' searchability to the same state in one call.

    Attributes:
        chunk_ids (list[str]): The chunks to toggle (at least one).
        enabled (bool): The state to apply to every listed chunk.
    """

    chunk_ids: list[str] = Field(min_length=1, description="The chunks to toggle (at least one).")
    enabled: bool = Field(description="The state to apply to every listed chunk.")


class ChunkEnabledResult(BaseModel):
    """
    The outcome of toggling one chunk's searchability.

    Attributes:
        chunk_id (str): The toggled chunk.
        enabled (bool): The recomputed EFFECTIVE state (override ?? role default).
        reindex_required (bool): True when a never-embedded chunk was enabled — needs a re-embed.
    """

    chunk_id: str = Field(description="The toggled chunk's UUID.")
    enabled: bool = Field(description="The recomputed effective searchability state.")
    reindex_required: bool = Field(
        description="True when a never-embedded chunk was enabled — needs a deferred re-embed."
    )


class BulkChunkEnabledResponse(BaseModel):
    """
    The per-chunk outcomes of a bulk toggle plus the ids that did not resolve to a chunk.

    Attributes:
        results (list[ChunkEnabledResult]): One outcome per KNOWN chunk.
        not_found (list[str]): Requested ids with no matching chunk (skipped, not an error).
    """

    results: list[ChunkEnabledResult] = Field(
        default_factory=list, description="One outcome per known chunk."
    )
    not_found: list[str] = Field(
        default_factory=list, description="Requested ids with no matching chunk."
    )


__all__ = [
    "MetadataValue",
    "DocumentListItem",
    "DocumentDetail",
    "PageInfo",
    "ChunkInfo",
    "ChunkEnabledPatch",
    "BulkChunkEnabledPatch",
    "ChunkEnabledResult",
    "BulkChunkEnabledResponse",
]

File: src/docforge/app/frontend/src/api/explorer.ts
// ====== Code Summary ======
// TypeScript mirror of the document explorer REST contract (catalogue, facts/metadata, pages,
// IR, chunks, blobs) + its typed client. Shapes copied verbatim from the backend's Pydantic
// models (see /openapi.json) — nothing invented.

import type { FieldOrigin } from "./collections";
import { apiFetch, apiFetchBlob, jsonInit } from "./http";
import type { JobEvent } from "./jobs";

const DOCUMENTS_BASE = "/api/v1/documents";
const COLLECTIONS_BASE = "/api/v1/collections";
const BLOBS_BASE = "/api/v1/blobs";
const CHUNKS_BASE = "/api/v1/chunks";

export type DocumentStatus = "pending" | "processing" | "done" | "failed" | "cancelled";
export type SourceKind = "digital_born" | "scanned" | "mixed";
export type EnrichmentKind = "classify" | "ocr" | "vlm" | "chart_to_data" | "table_summary";
export type EnrichmentStatus = "ok" | "failed" | "skipped";

/**
 * A chunk's structural classification, assigned by the pipeline (not a user choice). Drives the
 * default effective-enabled state: only `body` is searchable by default — every other role
 * defaults off and needs an explicit enable to become searchable.
 */
export type ChunkRole = "body" | "header_footer" | "toc" | "boilerplate";

/** One row of a collection's document catalogue (the browse list). */
export interface DocumentListItem {
  id: string;
  filename: string;
  format: string;
  status: DocumentStatus;
  page_count: number | null;
  file_size: number;
  created_at: string | null;
  title: string;
  language: string | null;
  /** The document-level searchability toggle — disabling hides every chunk regardless of role. */
  enabled: boolean;
}

/** One resolved metadata value — the schema field name, its stored value, and who filled it. */
export interface MetadataValue {
  field_name: string;
  value: unknown;
  origin: FieldOrigin;
}

/** A document's full facts plus its resolved document-level metadata. */
export interface DocumentDetail {
  id: string;
  collection_id: string;
  filename: string;
  format: string;
  mime_type: string;
  file_size: number;
  page_count: number | null;
  language: string | null;
  title: string;
  source_kind: SourceKind;
  status: DocumentStatus;
  source_hash: string;
  pdf_blob_hash: string | null;
  simhash: string | null;
  pipeline_version: string;
  created_at: string | null;
  metadata: MetadataValue[];
  /** The document-level searchability toggle — disabling hides every chunk regardless of role. */
  enabled: boolean;
}

/** One page's geometry, routing and its render blob reference. */
export interface PageInfo {
  page_number: number;
  width: number | null;
  height: number | null;
  is_scanned: boolean;
  language: string | null;
  render_blob_hash: string | null;
}

/** One raw IR block — structure, position in reading order and native text. */
export interface IRBlock {
  id: string;
  block_type: string;
  page: number;
  bbox: number[];
  reading_order: number;
  parent_id: string | null;
  level: number | null;
  text: string | null;
  is_boilerplate: boolean;
  language: string | null;
}

/** A TABLE block's parsed structure (1:1 with its block). */
export interface IRTable {
  block_id: string;
  n_rows: number;
  n_cols: number;
  has_header: boolean;
  /** Row-major 2D cell grid — untyped on the wire (JSON), narrowed defensively when rendered. */
  cells: unknown;
  linearized_md: string | null;
}

/** A FIGURE block's parse-time facts (1:1 with its block). */
export interface IRFigure {
  block_id: string;
  crop_blob_hash: string | null;
  caption_block_id: string | null;
}

/** One enrichment applied to a block (OCR/VLM/classify/chart_to_data/table_summary). */
export interface IREnrichment {
  id: string;
  block_id: string;
  kind: EnrichmentKind;
  text: string | null;
  data: unknown;
  status: EnrichmentStatus;
}

/** The full IR of a document — raw blocks, table/figure details and every enrichment. */
export interface DocumentIR {
  blocks: IRBlock[];
  tables: IRTable[];
  figures: IRFigure[];
  enrichments: IREnrichment[];
}

/** One retrieval chunk — enriched text, composition (block ids) and generated metadata. */
export interface ChunkInfo {
  id: string;
  chunk_index: number;
  text: string;
  token_count: number;
  is_indexed: boolean;
  strategy: string;
  parent_id: string | null;
  block_ids: string[];
  metadata: MetadataValue[];
  /** Structural classification assigned by the pipeline — drives the default enabled state. */
  role: ChunkRole;
  /** The recomputed EFFECTIVE searchability state (enabled_override ?? role_default_enabled). */
  enabled: boolean;
  /** The chunk's section breadcrumb (outer→inner headings); [] when none. */
  heading_path: string[];
  /** Page of the chunk's primary (leading) block; null when it has no located block. */
  page: number | null;
}

/** The desired searchability state for one chunk (its enabled_override). */
export interface ChunkEnabledPatch {
  enabled: boolean;
}

/**
 * The outcome of toggling one chunk's searchability.
 *
 * `reindex_required` is true only when enabling a chunk that was never embedded — it has no
 * Qdrant point, so it is NOT searchable until a later on-demand re-embed runs.
 *
 * `search_sync_pending`/`search_sync_error` (caught missing by `_contractParity.ts`, 2026-09):
 * meaningful on the single-chunk route — a bulk response's per-result entries always report
 * `false` here, the REQUEST-level flag on `BulkChunkEnabledResponse` is authoritative for a batch.
 */
export interface ChunkEnabledResult {
  chunk_id: string;
  enabled: boolean;
  reindex_required: boolean;
  search_sync_pending: boolean;
  search_sync_error: string | null;
}

/** Toggle several chunks' searchability to the same state in one call (the UI's multi-select). */
export interface BulkChunkEnabledPatch {
  chunk_ids: string[];
  enabled: boolean;
}

/**
 * The per-chunk outcomes of a bulk toggle plus the ids that did not resolve to a chunk.
 *
 * `search_sync_pending`/`search_sync_error` (caught missing by `_contractParity.ts`, 2026-09):
 * true when Postgres committed every toggle but the Qdrant payload sync failed — the search store
 * is stale until a re-run/backfill reconciles it. This is the request-level truth for the whole
 * batch (never a false full-success); the same fields on each nested `ChunkEnabledResult` stay
 * `false`/`null`.
 */
export interface BulkChunkEnabledResponse {
  results: ChunkEnabledResult[];
  not_found: string[];
  search_sync_pending: boolean;
  search_sync_error: string | null;
}

export function listDocuments(collectionId: string): Promise<DocumentListItem[]> {
  return apiFetch(`${COLLECTIONS_BASE}/${collectionId}/documents`);
}

export function getDocument(id: string): Promise<DocumentDetail> {
  return apiFetch(`${DOCUMENTS_BASE}/${id}`);
}

export function getDocumentPages(id: string): Promise<PageInfo[]> {
  return apiFetch(`${DOCUMENTS_BASE}/${id}/pages`);
}

export function getDocumentIR(id: string): Promise<DocumentIR> {
  return apiFetch(`${DOCUMENTS_BASE}/${id}/ir`);
}

/** A document's ingestion provenance — the parser/model pipeline that produced its IR + chunks. */
export interface DocumentProvenance {
  document_id: string;
  pipeline_version: string;
  /** The last ingestion job id — null when it has expired/been reaped (stages then empty). */
  job_id: string | null;
  /** False when no ingestion job survives (its stage timeline could not be recovered). */
  available: boolean;
  /** The per-stage trace in execution order: which node/parser/model ran, timing, token/cost. */
  stages: JobEvent[];
}

export function getDocumentProvenance(id: string): Promise<DocumentProvenance> {
  return apiFetch(`${DOCUMENTS_BASE}/${id}/provenance`);
}

export function getDocumentChunks(id: string): Promise<ChunkInfo[]> {
  return apiFetch(`${DOCUMENTS_BASE}/${id}/chunks`);
}

export function deleteDocument(id: string): Promise<void> {
  return apiFetch(`${DOCUMENTS_BASE}/${id}`, { method: "DELETE" });
}

/** Toggle one chunk's searchability (reversible, no re-embed). */
export function setChunkEnabled(id: string, enabled: boolean): Promise<ChunkEnabledResult> {
  return apiFetch(`${CHUNKS_BASE}/${id}/enabled`, jsonInit("PATCH", { enabled }));
}

/** Toggle several chunks' searchability to the same state in one call (the multi-select bar). */
export function setChunksEnabled(chunkIds: string[], enabled: boolean): Promise<BulkChunkEnabledResponse> {
  return apiFetch(`${CHUNKS_BASE}/enabled`, jsonInit("PATCH", { chunk_ids: chunkIds, enabled }));
}

/** The blob route for a content hash. Fetch it through `apiFetchBlob` (bearer) — NOT as a raw
 *  `<img src>`/`<a href>`, which a browser navigation can't authenticate. See `components/BlobImage`. */
export function blobUrl(hash: string): string {
  return `${BLOBS_BASE}/${hash}`;
}

// ====== On-the-fly markdown/HTML views (generated from the canonical IR, never a stored blob) ======
// Mirrors GET /documents/{id}/markdown and /documents/{id}/html. Unlike blobUrl above, these are
// NOT content-hash routes — the body is rendered fresh from the document's IR on every request.

export type DocumentViewFormat = "markdown" | "html";

function documentViewUrl(id: string, format: DocumentViewFormat, download: boolean): string {
  return `${DOCUMENTS_BASE}/${id}/${format === "markdown" ? "markdown" : "html"}?download=${download}`;
}

/** The `<stem>.md`/`<stem>.html` filename the backend would attach — mirrored here so a caller can
 *  name the client-side download (Content-Disposition on a `fetch()`-constructed blob is not
 *  honoured by the browser; only a navigation respects it, so we must set it ourselves). */
export function documentViewFilename(filename: string, format: DocumentViewFormat): string {
  const stem = filename.replace(/\.[^./]+$/, "") || "document";
  return `${stem}.${format === "markdown" ? "md" : "html"}`;
}

/** Fetch a document's markdown/HTML view (authenticated) and open it inline in a new tab — mirrors
 *  `api/blobs.ts`'s `openBlobInNewTab`, generalized to a plain route instead of a content-hash one. */
export async function openDocumentViewInNewTab(id: string, format: DocumentViewFormat): Promise<void> {
  const blob = await apiFetchBlob(documentViewUrl(id, format, false));
  const url = URL.createObjectURL(blob);
  window.open(url, "_blank", "noopener");
  window.setTimeout(() => URL.revokeObjectURL(url), 60_000);
}

/** Fetch a document's markdown/HTML view (authenticated) and trigger a browser download under `filename`. */
export async function downloadDocumentView(id: string, format: DocumentViewFormat, filename: string): Promise<void> {
  const blob = await apiFetchBlob(documentViewUrl(id, format, true));
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = filename;
  document.body.appendChild(anchor);
  anchor.click();
  anchor.remove();
  URL.revokeObjectURL(url);
}
